cdap-common/src/main/java/co/cask/cdap/common/zookeeper/ReentrantDistributedLock.java

```diff
@@ -0,0 +1,395 @@
+
+        // If the lock is completed (mainly due to cancellation), simply abort the lock acquisition logic.
+        if (completion.isDone()) {
+          return;
+        }
+
+        // Step 2-5. Try to determine who is the lock owner and watch for ZK node changes if itself is not the owner.
+        doAcquire(completion, waitForLock, guid, lockNode);
+      }
+
+      @Override
+      public void onFailure(Throwable t) {
+        if (t instanceof KeeperException.ConnectionLossException) {
+          // Ignore connection exception in create. Going to handle it in next step.
+          // See the ZK receipt for details about the possible failure situation that can cause this.
+          doAcquire(completion, waitForLock, guid, null);
+        } else {
+          LOG.error("Exception raised when creating lock node at {}", lockPath, t);
+          completion.setException(t);
+        }
+      }
+    });
+
+    // Gets the result from the completion
+    try {
+      if (interruptible) {
+        localLockNode.set(completion.get(timeout, unit));
+      } else {
+        localLockNode.set(Uninterruptibles.getUninterruptibly(completion, timeout, unit));
+      }
+      return true;
+    } catch (InterruptedException e) {
+      completion.cancel(true);
+      throw e;
+    } catch (TimeoutException e) {
+      completion.cancel(true);
+      throw e;
+    } catch (CancellationException e) {
+      // If the completion get cancelled, meaning the lock acquisition is aborted.
+      return false;
+    }
+  }
+
+  /**
+   * Executes the lock acquisition process. This corresponds to step 2-5 of the distributed lock receipt.
+   */
+  private void doAcquire(final SettableFuture<String> completion, final boolean waitForLock,
+                         final String guid, @Nullable final String lockPath) {
+    // Step 2. Get all children under the lock parent.
+    Futures.addCallback(zkClient.getChildren(path), new FutureCallback<NodeChildren>() {
+
+      @Override
+      public void onSuccess(NodeChildren children) {
+        // Find the lock node in case the creation step failed by matching the guid
+        // See "Recoverable Errors and the GUID" in the ZooKeeper guide
+        final String lockNode = lockPath == null ? findLockNode(children.getChildren(), guid) : lockPath;
+        if (lockNode == null) {
+          // If not able to find the lock node, fail the locking procedure.
+          completion.setException(new IllegalStateException("Failed to acquire lock").fillInStackTrace());
+          return;
+        }
+
+        if (lockPath == null) {
+          // If lock node was not determined in step 1 due to connection loss exception, need to add the
+          // node deletion handler in here after the actual lockNode is determined.
+          deleteNodeOnFailure(completion, lockNode);
+        }
+
+        // Find the node to watch, which is the one with the largest id that is smaller than currentId
```

Comment:
thanks for the comment. It is much easier to understand than the referred zookeeper recipe